File: web/src/lib/supabase/storage.ts
/**
 * Supabase Storage Helpers
 *
 * Server-side utility functions for reading and managing paper assets
 * stored in the private Supabase Storage "papers" bucket.
 *
 * Responsibilities:
 * - Generate signed URLs for thumbnails and figure images (private bucket)
 * - Download paper content (markdown, sections, figures metadata, metadata) via Supabase client
 * - Delete all storage assets for a paper (requires service role key)
 */

import { createClient as createSupabaseClient } from '@supabase/supabase-js';

// ============================================================================
// CONSTANTS
// ============================================================================

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL!;
const BUCKET_NAME = 'papers';
const SIGNED_URL_EXPIRY_SECONDS = 3600;

// ============================================================================
// TYPES
// ============================================================================

/** Content downloaded from storage for a single paper. */
export interface StoredPaperContent {
  finalMarkdown: string;
  sections: Record<string, unknown>[];
  figures: Record<string, unknown>[];
  metadata: Record<string, unknown>;
}

// ============================================================================
// SERVICE CLIENT
// ============================================================================

/**
 * Create a Supabase client using the service role key.
 * Required for all private bucket operations (signed URLs, downloads, deletes).
 * @returns Supabase client with service role permissions
 */
function getServiceClient() {
  const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY;
  if (!serviceRoleKey) {
    throw new Error('SUPABASE_SERVICE_ROLE_KEY is not set -- cannot access private storage bucket');
  }
  return createSupabaseClient(SUPABASE_URL, serviceRoleKey);
}

// ============================================================================
// SIGNED URL HELPERS
// ============================================================================

/**
 * Generate a signed URL for a paper's thumbnail image.
 * Returns null if the thumbnail doesn't exist in storage.
 * @param paperUuid - Unique identifier for the paper
 * @returns Signed HTTPS URL for thumbnail.png (valid for 1 hour), or null
 */
export async function getPaperThumbnailUrl(paperUuid: string): Promise<string | null> {
  const supabase = getServiceClient();
  const path = `${paperUuid}/thumbnail.png`;

  const { data, error } = await supabase.storage
    .from(BUCKET_NAME)
    .createSignedUrl(path, SIGNED_URL_EXPIRY_SECONDS);

  if (error || !data?.signedUrl) {
    return null;
  }

  return data.signedUrl;
}

/**
 * Generate signed thumbnail URLs for multiple papers in a single API call.
 * Returns a Map of paperUuid -> signedUrl (null for missing thumbnails).
 * @param paperUuids - Array of paper UUIDs
 * @returns Map from paperUuid to signed URL or null
 */
export async function getPaperThumbnailUrls(
  paperUuids: string[]
): Promise<Map<string, string | null>> {
  if (paperUuids.length === 0) return new Map();

  const supabase = getServiceClient();
  const paths = paperUuids.map((uuid) => `${uuid}/thumbnail.png`);

  const { data, error } = await supabase.storage
    .from(BUCKET_NAME)
    .createSignedUrls(paths, SIGNED_URL_EXPIRY_SECONDS);

  const result = new Map<string, string | null>();

  if (error || !data) {
    for (const uuid of paperUuids) result.set(uuid, null);
    return result;
  }

  for (let i = 0; i < paperUuids.length; i++) {
    const entry = data[i];
    result.set(paperUuids[i], entry?.error ? null : (entry?.signedUrl ?? null));
  }

  return result;
}

/**
 * Generate a signed URL for a specific figure image.
 * @param paperUuid - Unique identifier for the paper
 * @param figureId - The figure identifier (filename without extension)
 * @returns Signed HTTPS URL for the figure PNG (valid for 1 hour)
 */
export async function getPaperFigureUrl(paperUuid: string, figureId: string): Promise<string> {
  const supabase = getServiceClient();
  const path = `${paperUuid}/figures/${figureId}.png`;

  const { data, error } = await supabase.storage
    .from(BUCKET_NAME)
    .createSignedUrl(path, SIGNED_URL_EXPIRY_SECONDS);

  if (error || !data?.signedUrl) {
    throw new Error(`Failed to generate signed URL for ${path}: ${error?.message ?? 'no signed URL returned'}`);
  }

  return data.signedUrl;
}

// ============================================================================
// DOWNLOAD HELPERS
// ============================================================================

/**
 * Download all text/JSON content for a paper from storage.
 * Checks pipeline_version from metadata.json first:
 * - v2 papers (PDF-direct pipeline): returns empty content fields (no content.md/sections/figures)
 * - v1 papers: downloads content.md, sections.json, figures.json as before
 * @param paperUuid - Unique identifier for the paper
 * @returns StoredPaperContent with all text content fields populated
 */
export async function downloadPaperContent(paperUuid: string): Promise<StoredPaperContent> {
  const supabase = getServiceClient();
  const bucket = supabase.storage.from(BUCKET_NAME);
  const prefix = paperUuid;

  // Download metadata first to check pipeline version
  const metadataResult = await bucket.download(`${prefix}/metadata.json`);
  if (metadataResult.error || !metadataResult.data) {
    throw new Error(`Failed to download metadata.json for paper ${paperUuid}: ${metadataResult.error?.message}`);
  }

  const metadataText = await metadataResult.data.text();
  const metadata = JSON.parse(metadataText) as Record<string, unknown>;

  // v2 papers have no content.md, sections.json, or figures.json
  if (metadata.pipeline_version === 2) {
    return {
      finalMarkdown: '',
      sections: [],
      figures: [],
      metadata,
    };
  }

  // v1 pipeline: download remaining files
  const [markdownResult, sectionsResult, figuresResult] = await Promise.all([
    bucket.download(`${prefix}/content.md`),
    bucket.download(`${prefix}/sections.json`),
    bucket.download(`${prefix}/figures.json`),
  ]);

  if (markdownResult.error || !markdownResult.data) {
    throw new Error(`Failed to download content.md for paper ${paperUuid}: ${markdownResult.error?.message}`);
  }
  if (sectionsResult.error || !sectionsResult.data) {
    throw new Error(`Failed to download sections.json for paper ${paperUuid}: ${sectionsResult.error?.message}`);
  }
  if (figuresResult.error || !figuresResult.data) {
    throw new Error(`Failed to download figures.json for paper ${paperUuid}: ${figuresResult.error?.message}`);
  }

  const [finalMarkdown, sectionsText, figuresText] = await Promise.all([
    markdownResult.data.text(),
    sectionsResult.data.text(),
    figuresResult.data.text(),
  ]);

  return {
    finalMarkdown,
    sections: JSON.parse(sectionsText) as Record<string, unknown>[],
    figures: JSON.parse(figuresText) as Record<string, unknown>[],
    metadata,
  };
}

/**
 * Download only the markdown content for a paper.
 * Checks metadata.json for pipeline_version first:
 * - v2 papers: returns empty string (no content.md exists)
 * - v1 papers: downloads and returns content.md
 * @param paperUuid - Unique identifier for the paper
 * @returns The raw markdown string from content.md, or empty string for v2 papers
 */
export async function downloadPaperMarkdown(paperUuid: string): Promise<string> {
  const supabase = getServiceClient();
  const bucket = supabase.storage.from(BUCKET_NAME);

  // Check pipeline version from metadata
  const metadataResult = await bucket.download(`${paperUuid}/metadata.json`);
  if (metadataResult.data) {
    const metadataText = await metadataResult.data.text();
    const metadata = JSON.parse(metadataText) as Record<string, unknown>;
    if (metadata.pipeline_version === 2) {
      return '';
    }
  }

  // v1 pipeline: download content.md
  const { data, error } = await bucket.download(`${paperUuid}/content.md`);

  if (error || !data) {
    throw new Error(`Failed to download content.md for paper ${paperUuid}: ${error?.message}`);
  }

  return data.text();
}

// ============================================================================
// DELETE HELPERS
// ============================================================================

/**
 * Delete all storage files for a paper.
 * Uses a service-role Supabase client since delete requires elevated permissions.
 * Handles both v1 (content.md, figures, sections) and v2 (thumbnail + metadata only) formats.
 * @param paperUuid - Unique identifier for the paper
 */
export async function deletePaperAssets(paperUuid: string): Promise<void> {
  const supabase = getServiceClient();
  const bucket = supabase.storage.from(BUCKET_NAME);
  const prefix = paperUuid;

  // Include all possible file paths from both v1 and v2 formats
  const pathsToDelete = [
    `${prefix}/thumbnail.png`,
    `${prefix}/metadata.json`,
    `${prefix}/content.md`,
    `${prefix}/sections.json`,
    `${prefix}/figures.json`,
  ];

  // Try to read figures.json to find individual figure files (v1 only)
  try {
    const { data: figuresBlob, error: figuresError } = await bucket.download(`${prefix}/figures.json`);
    if (!figuresError && figuresBlob) {
      const figuresText = await figuresBlob.text();
      const figuresList = JSON.parse(figuresText) as Record<string, unknown>[];
      for (const fig of figuresList) {
        const identifier = fig.identifier as string | undefined;
        if (identifier) {
          pathsToDelete.push(`${prefix}/figures/${identifier}.png`);
        }
      }
    }
  } catch {
    // If figures.json doesn't exist or is malformed, skip figure path discovery.
  }

  const { error } = await bucket.remove(pathsToDelete);
  if (error) {
    throw new Error(`Failed to delete storage assets for paper ${paperUuid}: ${error.message}`);
  }
}

File: worker/papers/storage.py
"""
Supabase Storage client wrapper for paper assets.

Handles uploading, downloading, and deleting paper files in a private Supabase Storage bucket.

Responsibilities:
- Upload paper assets after processing (thumbnail, metadata with pipeline_version)
- Download paper content (handles both v1 OCR format and v2 PDF-direct format)
- Generate signed URLs for thumbnails and figures (private bucket)
- Delete all assets for a paper (handles both old and new format)
- Lazy-initialize the Supabase client from config
"""

import json
import logging
from dataclasses import dataclass
from typing import Dict, List, Optional

from supabase import create_client, Client

from shared.config import settings

logger = logging.getLogger(__name__)


# ============================================================================
# CONSTANTS
# ============================================================================

BUCKET_NAME = "papers"
SIGNED_URL_EXPIRY_SECONDS = 3600


# ============================================================================
# TYPES
# ============================================================================

@dataclass
class StoredPaperContent:
    """Content downloaded from storage for a single paper."""

    final_markdown: str
    sections: List[Dict]
    figures: List[Dict]
    metadata: dict


# ============================================================================
# CLIENT INITIALIZATION
# ============================================================================

_client: Optional[Client] = None


def _get_client() -> Client:
    """Return the lazily-initialized Supabase client.

    Returns:
        Supabase client instance configured with service key.

    Raises:
        RuntimeError: If SUPABASE_URL or SUPABASE_SERVICE_KEY are not set.
    """
    global _client
    if _client is None:
        if not settings.SUPABASE_URL or not settings.SUPABASE_SERVICE_KEY:
            raise RuntimeError(
                "SUPABASE_URL and SUPABASE_SERVICE_KEY must be set in environment"
            )
        _client = create_client(settings.SUPABASE_URL, settings.SUPABASE_SERVICE_KEY)
    return _client


def _bucket():
    """Return the storage bucket handle.

    Returns:
        Supabase storage bucket reference for the papers bucket.
    """
    return _get_client().storage.from_(BUCKET_NAME)


# ============================================================================
# MAIN FUNCTIONS
# ============================================================================

def upload_paper_assets(
    paper_uuid: str,
    thumbnail_bytes: bytes,
    metadata: dict,
) -> None:
    """Upload processed paper assets to Supabase Storage.

    Uploads thumbnail.png and metadata.json (with pipeline_version: 2) under {paper_uuid}/.

    Args:
        paper_uuid: Unique identifier for the paper.
        thumbnail_bytes: Raw PNG bytes for the thumbnail (NOT base64).
        metadata: Dict of usage summary, processing stats, cost info.
    """
    bucket = _bucket()
    prefix = paper_uuid

    # Add pipeline version to metadata
    metadata["pipeline_version"] = 2

    # Upload thumbnail
    bucket.upload(
        f"{prefix}/thumbnail.png",
        thumbnail_bytes,
        {"content-type": "image/png", "upsert": "true"},
    )

    # Upload metadata JSON
    bucket.upload(
        f"{prefix}/metadata.json",
        json.dumps(metadata, ensure_ascii=False).encode("utf-8"),
        {"content-type": "application/json", "upsert": "true"},
    )


def download_paper_content(paper_uuid: str) -> StoredPaperContent:
    """Download content for a paper from storage.

    Checks pipeline_version from metadata.json first:
    - v2 papers: return empty final_markdown, empty sections/figures (they don't exist in storage)
    - v1 papers (or no version): download content.md, sections.json, figures.json as before

    Args:
        paper_uuid: Unique identifier for the paper.

    Returns:
        StoredPaperContent with all text content fields populated.
    """
    bucket = _bucket()
    prefix = paper_uuid

    # Always download metadata first to check pipeline version
    metadata_bytes = bucket.download(f"{prefix}/metadata.json")
    metadata = json.loads(metadata_bytes.decode("utf-8"))

    pipeline_version = metadata.get("pipeline_version", 1)

    # v2 papers: no content.md, sections.json, or figures.json in storage
    if pipeline_version >= 2:
        return StoredPaperContent(
            final_markdown="",
            sections=[],
            figures=[],
            metadata=metadata,
        )

    # v1 papers: download all legacy files
    markdown_bytes = bucket.download(f"{prefix}/content.md")
    sections_bytes = bucket.download(f"{prefix}/sections.json")
    figures_bytes = bucket.download(f"{prefix}/figures.json")

    return StoredPaperContent(
        final_markdown=markdown_bytes.decode("utf-8"),
        sections=json.loads(sections_bytes.decode("utf-8")),
        figures=json.loads(figures_bytes.decode("utf-8")),
        metadata=metadata,
    )


def download_markdown(paper_uuid: str) -> str:
    """Download only the markdown content for a paper.

    Checks metadata for pipeline_version first. Returns empty string for v2 papers.

    Args:
        paper_uuid: Unique identifier for the paper.

    Returns:
        The raw markdown string from content.md, or empty string for v2 papers.
    """
    bucket = _bucket()
    prefix = paper_uuid

    # Check pipeline version from metadata
    try:
        metadata_bytes = bucket.download(f"{prefix}/metadata.json")
        metadata = json.loads(metadata_bytes.decode("utf-8"))
        if metadata.get("pipeline_version", 1) >= 2:
            return ""
    except Exception:
        # If metadata doesn't exist, try legacy download
        pass

    markdown_bytes = bucket.download(f"{paper_uuid}/content.md")
    return markdown_bytes.decode("utf-8")


def get_thumbnail_url(paper_uuid: str) -> str:
    """Generate a signed URL for a paper's thumbnail.

    Args:
        paper_uuid: Unique identifier for the paper.

    Returns:
        Signed HTTPS URL for thumbnail.png (valid for 1 hour).

    Raises:
        RuntimeError: If signed URL generation fails.
    """
    path = f"{paper_uuid}/thumbnail.png"
    response = _bucket().create_signed_url(path, SIGNED_URL_EXPIRY_SECONDS)
    signed_url = response.get("signedURL")
    if not signed_url:
        raise RuntimeError(
            f"Failed to generate signed URL for {path}: {response}"
        )
    return signed_url


def get_figure_url(paper_uuid: str, figure_id: str) -> str:
    """Generate a signed URL for a specific figure image.

    Args:
        paper_uuid: Unique identifier for the paper.
        figure_id: The figure identifier (used as filename without extension).

    Returns:
        Signed HTTPS URL for the figure PNG (valid for 1 hour).

    Raises:
        RuntimeError: If signed URL generation fails.
    """
    path = f"{paper_uuid}/figures/{figure_id}.png"
    response = _bucket().create_signed_url(path, SIGNED_URL_EXPIRY_SECONDS)
    signed_url = response.get("signedURL")
    if not signed_url:
        raise RuntimeError(
            f"Failed to generate signed URL for {path}: {response}"
        )
    return signed_url


def delete_paper_assets(paper_uuid: str) -> None:
    """Delete all storage files for a paper.

    Handles both v1 (OCR format with content.md, figures, sections) and
    v2 (just thumbnail + metadata). Includes all possible paths regardless
    of format -- the remove call handles non-existent files gracefully.

    Args:
        paper_uuid: Unique identifier for the paper.
    """
    bucket = _bucket()
    prefix = paper_uuid

    # Always include both old and new format paths
    paths_to_delete = [
        f"{prefix}/thumbnail.png",
        f"{prefix}/metadata.json",
        f"{prefix}/content.md",
        f"{prefix}/sections.json",
        f"{prefix}/figures.json",
    ]

    # Try to read figures.json to find individual figure files (v1 papers only)
    try:
        figures_bytes = bucket.download(f"{prefix}/figures.json")
        figures_list = json.loads(figures_bytes.decode("utf-8"))
        for fig in figures_list:
            identifier = fig.get("identifier")
            if identifier:
                paths_to_delete.append(f"{prefix}/figures/{identifier}.png")
    except Exception:
        # figures.json doesn't exist (v2 paper) or is malformed -- skip figure deletion
        pass

    bucket.remove(paths_to_delete)
